### PR TITLE
Feature/lattice 2892 collaboration api enhancements

### PR DESCRIPTION
**Pull request by solomontang**
- `getCollaborations` has been reworked to get Collaboration objects (READ) for each collaboration id provided
- `getAllCollaborations` gets all Collaboration objects the caller has READ permissions on (old getCollaborations behavior)
- `getCollaborationsWithDataSets` has been reworked to return Collaboration objects, instead of just their ids

### CODE REVIEW

File: datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt [hunk 1]
```diff
@@ -41,14 +43,24 @@ class CollaborationsController : AuthorizingComponent, CollaborationsApi {
...
         val authorizedCollaborationIds = getAllAuthorizedCollaborationIds()
         return collaborationService.getCollaborations(authorizedCollaborationIds).values
     }
 
+    @Timed
+    @GetMapping(
+            value = ["", "/"],
+            produces = [MediaType.APPLICATION_JSON_VALUE]
+    )
+    override fun getCollaborations(@RequestParam( ID_PARAM ) ids: Set<UUID>): Map<UUID, Collaboration> {
```
Comment (geekbeast) [inline datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt:59]: Do you want this in as the body or in the URL? That URL will get long real fast.

File: api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt [hunk 2]
```diff
@@ -23,6 +24,7 @@ interface CollaborationsApi {
         const val COLLABORATION_ID_PATH = "/{$COLLABORATION_ID_PARAM}"
         const val DATA_SET_ID_PARAM = "dataSetId"
         const val DATA_SET_ID_PATH = "/{$DATA_SET_ID_PARAM}"
+        const val ID_PARAM = "id"
```
Comment (UnsungHero97) [inline api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt:27]:
```suggestion
        const val IDS_PARAM = "ids"
```

File: api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt [hunk 3]
```diff
@@ -32,8 +34,16 @@ interface CollaborationsApi {
...
+    fun getAllCollaborations(): Iterable<Collaboration>
+
+    /**
+     * Gets [Collaboration] objects the caller has [Permission.READ] on with the provided ids.
+     *
+     * @return a Map<K, V> where K is a collaboration id and V a [Collaboration] object
+     */
     @GET(BASE)
-    fun getCollaborations(): Iterable<Collaboration>
+    fun getCollaborations(@Query(ID_PARAM) ids: Set<UUID>): Map<UUID, Collaboration>
```
Comment (UnsungHero97) [inline api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt:46]:
```suggestion
    fun getCollaborations(@Query(IDS_PARAM) ids: Set<UUID>): Map<UUID, Collaboration>
```

File: datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt (hunk 1 above)
Comment (UnsungHero97) [reply datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt:59]:
```suggestion
    override fun getCollaborations(@RequestParam( IDS_PARAM ) ids: Set<UUID>): Map<UUID, Collaboration> {
```

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt (hunk 2 above)
Comment (solomontang) [reply api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt:27]: the param should be id. if multiple ids are passed, they're each assigned their own in repeat
`?id=1234&id=4567`

File: datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt (hunk 1 above)
Comment (solomontang) [reply datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt:59]: using GET with a body, or implementing POST with body with the intention of using it as a GET both deviate from what is expect from REST.

length of url isn't really a concern. if we really run into character limitations, it should be handled by limiting API calls to a certain number of ids per request.

File: api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt (hunk 2 above)
Comment (UnsungHero97) [reply api/src/main/kotlin/com/openlattice/collaborations/CollaborationsApi.kt:27]: We discussed this in slack, agreeing to keep it as `id`.

**Review by UnsungHero97 — DISMISSED** [review]

File: datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt (hunk 1 above)
Comment (geekbeast) [reply datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt:59]: The comment about GET makes sense, though I am not as devoted to REST philosophy when (ab)using it for RPC. You'll have to deal with paging eventually so I think this is fine for now.

Comment (geekbeast) [reply datastore/src/main/kotlin/com/openlattice/collaborations/controllers/CollaborationsController.kt:59]: Also there is a trade off here for debugging/testing this API, but I know where those tickets will go :)

**Review by UnsungHero97 — APPROVED** [review]

**Review by geekbeast — APPROVED** [review]

**Pull request merged**